### PR TITLE
[WIP] Pass verbosity flag to ansible-inventory

### PR DESCRIPTION
**Pull request by AlanCoding**
Connect https://github.com/ansible/awx/issues/2105

See discussion at https://github.com/ansible/ansible/issues/50100

Let me recap the important point - either we need acceptance of this workaround on our side, or this issue needs to be punted back to the Ansible core side. One of these things needs to happen in order for the OpenStack plugin transition to happen.

It just happens to be that this verbosity issue can easily be resolved by the same workaround. So we get a nice extra little enhancement out of it too.

### CODE REVIEW

**Comment by AlanCoding** [comment]
recheck

File: awx/main/utils/ansible.py [hunk 1]
```diff
@@ -97,3 +97,72 @@ def could_be_inventory(project_path, dir_path, filename):
...
     return inventory_rel_path
+
+
+# This method is copied directly from Ansible core code base
+# lib/ansible/module_utils/json_utils.py
+# For purpose, see: https://github.com/ansible/ansible/issues/50100
+# Any patches to this method should sync from that version
+# NB: a copy of this function exists in ../../modules/core/async_wrapper.py. Ensure any
+# changes are propagated there.
+def _filter_non_json_lines(data):
```
Comment (ryanpetrello) [inline awx/main/utils/ansible.py:108]: No, I **really** don't want to do this :(

cc @ghjm @wenottingham we should apply pressure for core to fix this for us if we wanted it supported in 2.8

Comment (AlanCoding) [reply awx/main/utils/ansible.py:108]: I'm going to have to push back on this, and @matburt I know you are of the same mind here. Let me share an output with you:

https://gist.github.com/AlanCoding/a45892d6a45db098c7b30d6ab95621ea

This is a real error that people (other than me! @kdelee) are getting when running inventory imports in the branch for transitioning to inventory plugins.

> Failed to parse /tmp/awx_420_8nt12vmo/openstack.yml with                                                                                                                                            
openstack plugin: 'NoneType' object has no attribute 'set'

Okay! What object here was NoneType? Can I see a line where a `.set` is done on an object? Could I get lines from where that line was called? No. You may not. Not without this PR merged.

Let me stress, very strongly, that this was not a thing with scripts. Errors produced by the script plugin were not previously a consideration. The script itself errors all the time, but the plugin was only a thin layer to interact with the script. All that has changed. Errors will not occur in the plugin code all the time.

This is a directly user-facing problem. Errors that used to be shown as tracebacks will no longer be available unless we enable verbosity, and along with that we _have_ to trim the output. At all.

This is an architectural design problem. This was inherently there when the concept of `ansible-inventory` plus plugins was committed to.

We could figure out how to pass a file, and then have `ansible-inventory` pipe different sources of output to different handles. I totally agree a solution of that type could be very good, much better than this. But for one, that's going to take a long time to roll out. And another, we careful consideration for this seriously major cross-software design.

Comment (ryanpetrello) [reply awx/main/utils/ansible.py:108]: Isn't this function - based on my initial read - going to make inventory sources with _lots_ of verbose output (Satellite6 comes to mind which frequently outputs 100MB+ at higher verbosity levels)  **miserably** slow?

It literally loads it all into memory and line by line does string comparison on it.

I worry that the overhead of this function is going to have a _very_ drastic performance affect on inventory import speed for large imports.

Comment (ryanpetrello) [reply awx/main/utils/ansible.py:108]: At the _very_ least, can we limit this behavior to plugins that won't guarantee JSON output on stdout (OpenStack)?

Comment (AlanCoding) [reply awx/main/utils/ansible.py:108]: limiting the behavior was my intent with the wrapping function. Two simple checks of `endswith` and `startswith` shouldn't be duplicating the string, I don't think. But yes, that's a very very good point. I am kind of concerned about the case of _middle of the road_ verbosity with large outputs. The simple fact that verbosity is on dictates a necessity to do the expensive splitting, so that may need some guard rails. It's true that this is mainly useful on failed imports, and opposed to those that will be successful.... but success/failed state is hard to predict in advance.

Comment (ryanpetrello) [reply awx/main/utils/ansible.py]: At the end of the day, _in my opinion_, this is a failing of `ansible-inventory` as a tool.  If we can't rely on `ansible-inventory --list` (or perhaps some other CLI call) to **always, reliably** print valid JSON, we're going to keep dealing with this everywhere.

**Comment by ryanpetrello** [comment]
@AlanCoding given your recent contributions to the openstack inventory plugin, is it safe to safe we can close this PR?

**Comment by AlanCoding** [comment]
@ryanpetrello that addressed 1 problem, and yes, I am now clear to make a PR to implement openstack as a plugin, which I am working on.

We are blocked on a different issue, which is now articulated for our case here:

https://github.com/ansible/ansible/issues/52244

Let me emphasize the particular patter that was changed. We used to have a script like `cloudawesome.py` that did something like:

```python
import awesomecloudsdk
import json
print(json.dumps(awesomecloudsdk.get_me_my_hosts()))
```

We have to plan for the library `awesomecloudsdk` throwing exceptions, and giving meaningful information that we need to make decisions.

I had to go back _2 days_ in the issue queue to find an example.

https://github.com/ansible/awx/issues/3231

The transition to plugins removed this from an external script, and internalized it into code that Ansible runs within the process / python context (don't know if I have the best words for this), but just imported using their "plugin" utilities. That means that tracebacks are reported as Ansible tracebacks.

Once we implement plugins, a bug of the same type as 3231 will start to be reported as:

> 'dict' object has no attribute 'split'

And that's it. This is why we need to pass verbosity to `ansible-inventory`.

While I get that we can't go forward with this solution, this is a burning hot issue that needs fixing.

**Comment by ryanpetrello** [comment]
@AlanCoding I understand the _why_ of this PR given our options _today_.  If I haven't made myself abundantly clear at this point, it's my **very strong** opinion 😉 that `ansible-inventory` should solve this problem for us, and that we shouldn't have to care about this (because isn't the _entire_ point of our usage of that tool that it gives us valid JSON?)  Hopefully this problem goes away when https://github.com/ansible/ansible/issues/52244 is addressed and resolved.

cc @wenottingham @ghjm @thaumos I know this is on your radar, but I _really_ want to underscore that in my opinion, we _should not_ (maybe even **could not** - without notable performance regressions, assuming we go with the code in this PR) ship inventory plugin support in awx without https://github.com/ansible/ansible/issues/52244 being addressed in ansible itself.

**Comment by AlanCoding** [comment]
I'll close for now

In this diff, I count 5 line changes that we actually need for the more narrow task of "pass verbosity to ansible-inventory"

the rest of it encompasses the much more complex task of "and post-process stdout to strip off gunk"

**Comment by kdelee** [comment]
@AlanCoding @ryanpetrello if I am understanding correctly, we would like to be able to get access to the traceback in stderr WITHOUT passing the `-vvv` flag, because verbose output from some plugins can be horribly slow, but we want the traceback if something blows up, correct? I should add that to the ticket if so

**Comment by AlanCoding** [comment]
> because verbose output from some plugins can be horribly slow

Verbosity doesn't apply to the plugin. The plugin can throw an error, and maybe its client library should write to stderr before it bubbles up the exception. But if the import is _successful_, it should not write anything to stdout (this was the openstack bug).

There is no hook for the inventory plugin to even receive the Ansible verbosity level. AFAIK, all plugins are blissfully ignorant of this.

Verbosity only applies to Ansible core code. This imports and runs the plugins, but also catches _all_ exceptions from it.

The problem is that with verbosity at 0, it doesn't write the traceback. If you do any higher verbosity, it will write the traceback, but it will _also_ write a bunch of canned generic debugging messages about just general config and search path stuff.

> we would like to be able to get access to the traceback in stderr WITHOUT passing the `-vvv` flag

I don't care if we have to pass the `-vvv` flag.

There are multiple ways that core can go about resolving this, and almost all of them will require some change on our side to match. That's fine.

They could print tracebacks when on verbosity level of 0, and then don't even bother redirecting debug messages to stderr. If it's failing, then I don't care if the output is parsable, because we won't parse it. But I know enough of how they work to think this approach is unlikely. Plugins fail _every time_ on successful playbook runs.

Ansible views inventory as a list of many inventories, all of which are tested by every plugin, and then parsed (if testing passes via `verify_file`) by every plugin. In this view, a parse failure of _all_ plugins for 1 inventory is ignored, because hey, we have a plethora of inventories. Also, the `ini` plugin _loves_ to "claim" everything, returning True from `verify_file`, and then fail on the parsing step. This leads to _really_ confusing output, and some bugs that I have not yet carded because it's not urgent.

I'm sorry this is just a big tangent now.

bottom line: core changes something, which results in instructions for us, for how we can both get tracebacks, and not have stdout combing debug output with JSON. On our side, we are agnostic to what and how of that.

**Comment by kdelee** [comment]
I get you, I do find the output like this:
```
  File "/home/elijah/sfw/ansible/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/home/elijah/sfw/ansible/ansible/lib/ansible/plugins/inventory/yaml.py", line 112, in parse
    raise AnsibleParserError('Plugin configuration YAML file, not YAML inventory')

  File "/home/elijah/sfw/ansible/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/home/elijah/sfw/ansible/ansible/lib/ansible/plugins/inventory/constructed.py", line 99, in parse
    self._read_config_data(path)
  File "/home/elijah/sfw/ansible/ansible/lib/ansible/plugins/inventory/__init__.py", line 206, in _read_config_data
    raise AnsibleParserError("Incorrect plugin name in file: %s" % config.get('plugin', 'none found'))
```

as clear as mud :crying_cat_face:  seeing as I am using the ec2 plugin and I was even required to specify that I was using the ec2 plugin, but no! :expressionless: :boom:
